### PR TITLE
CIS compliance: logging

### LINKED ISSUE #187: Enable diagnostic logging for all resources + network
As per CIS 5. 

- [ ] Enable diagnostic logging with 'appropriate' metrics logged per resource
- [ ] Ensure logs are sent to log analytics
- [ ] Ensure log analytics storage account is private
- [ ] Retention greater than 90 days / exported to storage
- [ ] NSG Flow logs sent to log analytics
- [ ] Network Watcher enabled

### PR DESCRIPTION
**Pull request by t-young31**
# Resolves #187 

Adds NSG for vnet + network watcher logs if it's defined in the root config

### CODE REVIEW

File: infrastructure/core/main.tf [hunk 1]
```diff
@@ -100,17 +100,27 @@ resource "azurerm_log_analytics_workspace" "core" {
   resource_group_name        = azurerm_resource_group.core.name
   internet_ingestion_enabled = var.tf_in_automation ? false : true
   sku                        = "PerGB2018"
-  retention_in_days          = 30
+  retention_in_days          = 90
+  internet_query_enabled     = false
```
Comment (t-young31) [inline infrastructure/core/main.tf:104]: worth discussing: (1) do we want to disable this and (2) without an admin jumpbox how does how is it possible to access logs?

File: config.sample.yaml [hunk 2]
```diff
@@ -28,6 +28,10 @@ serve:
   github_app_id: __CHANGE_ME__
   github_app_installation_id: __CHANGE_ME__
 
-alert_recipients:  # Optional
-  - name: __CHANGE_ME__
-    email: __CHANGE_ME__  # e.g. employee@example.com
+monitoring:
```
Comment (jjgriff93) [inline config.sample.yaml:31]: ;)

File: infrastructure/core/main.tf (hunk 1 above)
Comment (jjgriff93) [reply infrastructure/core/main.tf:104]: I think this should be left enabled, would make monitoring very difficult. @damoodamoo thoughts?

File: infrastructure/core/main.tf [hunk 3]
```diff
@@ -120,7 +130,7 @@ resource "azurerm_monitor_action_group" "p0" {
 
   lifecycle {
     precondition {
```
Comment (jjgriff93) [inline infrastructure/core/main.tf:132]: Very nice!

File: infrastructure/core/network.tf [hunk 4]
```diff
@@ -77,3 +77,46 @@ resource "azurerm_private_dns_zone_virtual_network_link" "all" {
...
+
+# Network watcher must be added in real-data scenarios, thus will fail if network_watcher is undefined
+resource "azurerm_network_watcher_flow_log" "data_sources" {
+  count                     = (var.monitoring.network_watcher != null) || var.accesses_real_data ? 1 : 0
+  name                      = "nw-log-${var.naming_suffix}"
+  resource_group_name       = var.monitoring.network_watcher.resource_group_name
+  network_watcher_name      = var.monitoring.network_watcher.name
+  network_security_group_id = azurerm_network_security_group.core.id
+  storage_account_id        = azurerm_storage_account.core.id
+  enabled                   = true
```
Comment (jjgriff93) [inline infrastructure/core/network.tf:108]: Is this not the default? I.e. does it need to be specified?

File: infrastructure/core/network.tf [hunk 5]
```diff
@@ -77,3 +77,46 @@ resource "azurerm_private_dns_zone_virtual_network_link" "all" {
...
+    protocol                   = "*"
+    direction                  = "Outbound"
+    destination_address_prefix = "Internet"
+    destination_port_range     = 443
+    source_address_prefix      = "*"
+    source_port_range          = "*"
+  }
+}
+
+# Network watcher must be added in real-data scenarios, thus will fail if network_watcher is undefined
```
Comment (jjgriff93) [inline infrastructure/core/network.tf:100]: Worth a precondition for better failure handling if the config isn't set correctly?

File: infrastructure/core/network.tf [hunk 6]
```diff
@@ -77,3 +77,46 @@ resource "azurerm_private_dns_zone_virtual_network_link" "all" {
     azurerm_private_dns_zone.all
   ]
 }
+
+resource "azurerm_network_security_group" "core" {
```
Comment (jjgriff93) [inline infrastructure/core/network.tf:81]: Did databricks deploy ok in your testing? Nothing jumps out that would be blocked that databricks needs but worth us being sure

File: infrastructure/core/variables.tf [hunk 7]
```diff
@@ -99,11 +99,21 @@ variable "accesses_real_data" {
...
-  description = "List of recipients to receive alerts"
-  type = list(object({
-    name  = string
-    email = string
-  }))
-  default = []
+variable "monitoring" {
+  description = "Monitoring block"
+  type = object({
+    alert_recipients = list(object({ # List of recipients to receive alerts
```
Comment (jjgriff93) [inline infrastructure/core/variables.tf:105]: Should alert recipients be able to be set without setting network watcher and vice versa? If so, you'll need to mark these both as optional and provide individual defaults

File: config.sample.yaml (hunk 2 above)
Comment (t-young31) [reply config.sample.yaml:31]: 😉

File: infrastructure/core/network.tf (hunk 4 above)
Comment (t-young31) [reply infrastructure/core/network.tf:108]: it is, but explicit is good(?) and safeguards against the provider altering the default under us

File: infrastructure/core/network.tf (hunk 6 above)
Comment (t-young31) [reply infrastructure/core/network.tf:81]: let's wait and see 😅 https://github.com/UCLH-Foundry/FlowEHR/actions/runs/4552205929/jobs/8027474893

Comment (t-young31) [reply infrastructure/core/network.tf:81]: databricks creates its own NSG with a bunch of 100 ish priorities so as long as NSGs get merged we should(🤞🏼) be ok

Comment (t-young31) [reply infrastructure/core/network.tf:81]: `make test-transform` works locally

**Review by damoodamoo — APPROVED** [review]

File: infrastructure/core/main.tf (hunk 1 above)
Comment (t-young31) [reply infrastructure/core/main.tf]: discussed on slack – will disable in prod with a manual step to enable

**Comment by t-young31** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/4576552084 (with refid `11e81097`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/219#issuecomment-1492103119) from @t-young31)

**Comment by t-young31** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/4576567614 (with refid `11e81097`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/219#issuecomment-1492107171) from @t-young31)

**Comment by t-young31** [comment]
:robot: pr-bot :robot:

:runner: Running tests: https://github.com/UCLH-Foundry/FlowEHR/actions/runs/4576567614 (with refid `11e81097`)

(in response to [this comment](https://github.com/UCLH-Foundry/FlowEHR/pull/219#issuecomment-1492107171) from @t-young31)

**Pull request merged**